python/lib/core/dmod/core/common/collections/catalog.py

```diff
@@ -0,0 +1,651 @@
+        if isinstance(other, type(self.data)):
+            return self.data == hash_hashable_map_sequence(other)
+
+        if not isinstance(other, CatalogEntry):
+            return False
+
+        return self.data_hash == other.data_hash
+
+    def __hash__(self):
+        return self.data_hash
+
+    def __le__(self, other):
+        if isinstance(other, type(self.data)):
+            return self.data < other or self.data_hash == hash_hashable_map_sequence(other)
+
+        if not isinstance(other, CatalogEntry):
+            raise TypeError(f"Cannot perform {self.__class__.__name__} <= {other.__class__.__name__}")
+
+        return self.last_accessed <= other.last_accessed
+
+    def __lt__(self, other):
+        if isinstance(other, type(self.data)):
+            return self.data < other
+
+        if not isinstance(other, CatalogEntry):
+            raise TypeError(f"Cannot perform {type(self.data)} < {type(other)}")
+
+        return self.last_accessed < other.last_accessed
+
+    def __ge__(self, other):
+        if not isinstance(other, CatalogEntry):
+            raise TypeError(f"Cannot perform {self.__class__.__name__} >= {other.__class__.__name__}")
+
+        return self.last_accessed >= other.last_accessed
+
+    def __gt__(self, other):
+        if not isinstance(other, CatalogEntry):
+            raise TypeError(f"Cannot perform {self.__class__.__name__} < {other.__class__.__name__}")
+
+        return self.last_accessed >= other.last_accessed
+
+    def __str__(self):
+        return f"CacheEntry: {self.identifier} @ {self.last_accessed.strftime('%Y-%m-%d %H:%M:%S')}"
+
+    def __repr__(self):
+        return "{" + \
+            f'"identifier": {self.__data}, ' \
+            f'"data_hash": {self.__data_hash}, ' \
+            f'"last_accessed": "{self.last_accessed}"' + \
+            "}"
+
+
+# Part of Issue https://github.com/NOAA-OWP/DMOD/issues/434
+#   "Make metric computations asynchronous by location"
+class InputCatalog(EventfulMap[str, CatalogEntry[HashableTypeT]], typing.Generic[HashableTypeT]):
+    """
+    A base class that implements a caching mechanism that organizes entries based on access time and fires events
+    when adding, removing, accessing, and updating entries
+    """
+
+    def __init__(
+        self,
+        max_size: int = 100,
+        *,
+        values: typing.Mapping[str, typing.Union[HashableTypeT, CatalogEntry[HashableTypeT], None]] = None
+    ):
+        """
+        Constructor
+
+        Args:
+            max_size: The maximum number of items that this cache may contain. Values other than positive numbers will
+                cause the cache to become unbounded
+            values: Preexisting data to add to the cache
+            kwargs: Additional items to store in the cache
+        """
+        if values is None:
+            values = {}
+
+        super().__init__(contents=values)
+
+        self.__earliest: typing.Optional[CatalogEntry[HashableTypeT]] = None
```

Comment:
```suggestion
        self.__oldest: typing.Optional[CatalogEntry[HashableTypeT]] = None
```

I suggest calling this `oldest` instead of `earliest`. I think that is more intuitive. Thoughts?

Reply:
Re thinking this, I think it makes more sense to move this responsibility into something like an `lru_dict`.